File: frontend/src/lib/queries/assistant.ts
'use client'

import { useMutation } from '@tanstack/react-query'
import { useApi } from '@/lib/api/use-api'

export type AskMessage = { role: string; content: string }

export type AskResult = {
  answer: string
  output: string
  error: string | null
  figure: { data: any[]; layout: Record<string, any> } | null
  preview: Record<string, unknown>[] | null
}

export function useAskDaas(projectId: string) {
  const { call } = useApi()
  return useMutation({
    mutationFn: (input: { message: string; history: AskMessage[] }) =>
      call<AskResult>(`/api/v1/projects/${projectId}/ask`, { method: 'POST', json: input }),
  })
}

File: backend/app/schemas/assistant.py
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field


class AskMessage(BaseModel):
    role: str
    content: str


class AskRequest(BaseModel):
    message: str
    history: list[AskMessage] = Field(default_factory=list)
    model: str | None = None


class AskResponse(BaseModel):
    answer: str
    output: str = ""
    error: str | None = None
    figure: dict[str, Any] | None = None
    preview: list[dict[str, Any]] | None = None
